### PR TITLE
Remove `ClipRect` from Snackbar

### LINKED ISSUE #94701: Floating Snackbar with set width has clipped shadow
When a SnackBar is shown with the behavior set to floating and a width is specified, its shadow is noticeably clipped on the left and right sides.  My expectation is that the SnackBar should display with a natural shadow, as seen in the Material Design documentation.

Here's what you'll see if you run the provided code sample:

Screen Shot 2021-12-05 at 5 40 52 PM

Maybe this is related to #47202, but I'm not sure.

<details>
<summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(const SnackBarDemo());

class SnackBarDemo extends StatelessWidget {
  const SnackBarDemo({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Clipped SnackBar Shadow Demo',
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Clipped SnackBar Shadow Demo'),
        ),
        body: const SnackBarPage(),
      ),
    );
  }
}

class SnackBarPage extends StatelessWidget {
  const SnackBarPage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Center(
      child: ElevatedButton(
        onPressed: () {
          const snackBar = SnackBar(
            content: Text('My shadow is clipped :(', style: TextStyle(color: Colors.black,),),
            elevation: 12.0,
            width: 344, // Material Design spec
            behavior: SnackBarBehavior.floating,
            backgroundColor: Colors.white,
            duration: Duration(seconds: 10),
          );

          ScaffoldMessenger.of(context).showSnackBar(snackBar);
        },
        child: const Text('Show Sad SnackBar'),
      ),
    );
  }
}
```
</details>

<details>
  <summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel beta, 2.8.0-3.2.pre, on macOS 11.6 20G165 darwin-x64, locale en-US)
    • Flutter version 2.8.0-3.2.pre at /Users/bryson/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 2901cd7208 (2 weeks ago), 2021-11-18 11:15:09 -0800
    • Engine revision bcc2b7f12c
    • Dart version 2.15.0 (build 2.15.0-268.18.beta)

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/bryson/Library/Android/sdk
    • Platform android-31, build-tools 29.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.62.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.29.0

[✓] Connected device (2 available)
    • Pixel 6 Pro (mobile) • 192.168.1.198:42951 • android-arm64  • Android 12 (API 31)
    • Chrome (web)         • chrome              • web-javascript • Google Chrome 96.0.4664.55

• No issues found!
```

```
```

```
```

</details>

### PR DESCRIPTION
**Pull request by TahaTesser**
Fixes https://github.com/flutter/flutter/issues/94701

Complete details https://github.com/flutter/flutter/issues/94701#issuecomment-987885279

Based on this feedback, I notice a lot of inner `snackbar` widget instances already have `clip.none` as default, and I also tried the material widget with a container for `clip.none` yet the clipping was not going away then I realized, `ClipRect` is actually the widget that is doing this clipping here. Replacing this widget with a simple container solves the clipping but it does affect a few golden tests as expected.

`Golden "material.snack_bar.goldenTest.workWithBottomSheet.png": Pixel test failed, 0.16% diff
detected.`

| Before | After |
| --------------- | --------------- |
|  |  |

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by TahaTesser** [comment]
cc: @Piinks 

I can see ClipRect was added in https://github.com/flutter/flutter/pull/66504/files#diff-f28e065ae18b9a43c269b6b6e56ceb417084872226ea345fd3e3e207837f0f3dR563

Is it necessary for Hero transition? Is this a good change? If you have any alternative fix

**Comment by skia-gold** [comment]
Gold has detected about 6 new digest(s) on patchset 1.
View them at https://flutter-gold.skia.org/cl/github/94811

**Comment by Piinks** [comment]
I don't think it was necessary for the Hero, the ClipRect was already there before I made that change.

**Review by Piinks — COMMENTED** [review]
I am not sure why there was a ClipRect there to begin with, I would look in the git history to see if it was intentional so we don't introduce a regression.

There is a duplicate issue that is closed: 
- https://github.com/flutter/flutter/issues/47202

It was fixed by:
- https://github.com/flutter/flutter/pull/47616 (reverted)
- Relanded with migration in
  - https://github.com/flutter/flutter/pull/50597
  - https://github.com/flutter/flutter/pull/54985
  - https://github.com/flutter/flutter/pull/55064

So either we regressed already, or the bug was never fixed and was closed accidentally (those PRs all appear to be related to positioning). In either case, we should add more explicit tests so we know for sure if this comes up again in the future.

**Comment by TahaTesser** [comment]
@Piinks 
Thanks so much for the feedback, I just realized, it wasn't added but kept 

I already looked into https://github.com/flutter/flutter/issues/47202 and its PRs, it seems to be a related position indeed and it shows shadow being clipped at the bottom by navigation but the new issue is only reproducible when you force a width  and horizontal sides are clipped 
(this parameter wasn't available in that version of Fluttter where this was handled).



Here I created a basic code sample that doesn't use Snackbar at all, just to demonstrate what's happening with `ClipRect` and container without horizontal margin and this issue seems to beyond Snackbar since affect any widget with elevation 


Screenshot 2021-12-08 at 11 32 58


<details> 
<summary>minimal code sample</summary> 

```dart
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'Material App',
      home: Home(),
    );
  }
}

class Home extends StatelessWidget {
  const Home({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Material App Bar'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          children: [
            ClipRect(
              child: Container(
                margin: const EdgeInsets.only(bottom: 15, top: 10),
                width: 300,
                child: const Material(
                    elevation: 8,
                    shape: RoundedRectangleBorder(),
                    child: Padding(
                      padding: EdgeInsets.all(8.0),
                      child:
                          Text('ClipRect > Container(width: 300) > Material'),
                    )),
              ),
            ),
            Container(
              margin: const EdgeInsets.only(bottom: 15, top: 10),
              width: 300,
              child: const Material(
                  elevation: 8,
                  shape: RoundedRectangleBorder(),
                  child: Padding(
                    padding: EdgeInsets.all(8.0),
                    child: Text('Container(width: 300) > Material'),
                  )),
            ),
            ClipRect(
              child: Container(
                margin: const EdgeInsets.only(bottom: 15, top: 10, left: 15, right: 15),
                child: const Material(
                    elevation: 8,
                    shape: RoundedRectangleBorder(),
                    child: Padding(
                      padding: EdgeInsets.all(8.0),
                      child: Text('Container with horizontal margins > Material'),
                    )),
              ),
            ),
          ],
        ),
      ),
    );
  }
}

``` 
	
</details>

File: packages/flutter/lib/src/material/snack_bar.dart [hunk 1]
```diff
@@ -612,7 +612,7 @@ class _SnackBarState extends State<SnackBar> {
 
     return Hero(
       tag: '<SnackBar Hero tag - ${widget.content}>',
-      child: ClipRect(child: snackBarTransition),
+      child: Container(child: snackBarTransition),
```
Comment (Piinks) [inline packages/flutter/lib/src/material/snack_bar.dart:615]: Can we just remove the Container?

**Comment by skia-gold** [comment]
Gold has detected about 10 new digest(s) on patchset 2.
View them at https://flutter-gold.skia.org/cl/github/94811

**Comment by TahaTesser** [comment]
@Piinks 
Thanks for the suggestion, without container, there are still golden test diffs

**Review by Piinks — APPROVED** [review]
Thanks @TahaTesser! I am surprised the analyzer did not complain about an empty Container wrapper. 
This and the golden files LGTM! The bot does not recognize that those are tests, I'll reach out to @Hixie for exception. :)

**Comment by Hixie** [comment]
test-exempt: is tested (gold)

Looks like this also breaks some Google-internal tests, you'll need to ask someone to take a look at those to see if it's anything serious or just more golden tests. Ask around in #hackers-tests.

**Comment by Piinks** [comment]
Thanks @Hixie! I have taken a look at the Google-internal failures. They all look as expected, there is just a sliver of shadow that was previously being clipped incorrectly, similar to the golden file updates here.
Will manually set the check to green here and follow up when it rolls. 👍

**Pull request merged**